kernel/mmanutils.c

```diff
@@ -184,9 +267,134 @@ void* myst_mremap(
     return p;
 }
 
+MYST_UNUSED
+static void _dump_msync_mappings(void)
+{
+    for (msync_mapping_t* p = _msync_mappings; p; p = p->next)
+    {
+        printf("[%p][%zu][%zu]\n", p->addr, p->length, p->length / 4096);
+    }
+
+    printf("\n");
+}
+
+/* release msync mappings that are contained in the range [addr:addr+length] */
+static int _release_msync_mappings(void* addr, size_t length)
+{
+    int ret = 0;
+
+    myst_spin_lock(&_msync_mappings_lock);
+    {
+        msync_mapping_t* p = _msync_mappings;
+        msync_mapping_t* prev = NULL;
+
+        while (p)
+        {
+            msync_mapping_t* next = p->next;
+            uint8_t* lo = addr;
+            uint8_t* hi = (uint8_t*)addr + length;
+            uint8_t* plo = p->addr;
+            uint8_t* phi = (uint8_t*)p->addr + p->length;
+            uint8_t* maxlo = _max_ptr(lo, plo);
+            uint8_t* minhi = _min_ptr(hi, phi);
+
+            /* if there is an overlap */
+            if (maxlo < minhi)
+            {
+                size_t llength = maxlo - plo;
+                size_t rlength = phi - minhi;
+                size_t roffset = p->offset + (minhi - plo);
+
+                //     .........
+                // ..........
+                //
+                // .........
+                //     ..........
+
+                // left range:  [plo:llength]
+                // right range: [maxhi:rlength]
+
+                if (llength && rlength)
+                {
+                    // printf("case1: split\n");
+                    msync_mapping_t* rm;
+
+                    /* create the right mapping */
+                    if (!(rm = _new_msync_mapping(
+                              p->fd,
+                              roffset, /* offset */
+                              minhi,   /* addr */
+                              rlength))) /* length */
+                    {
+                        ERAISE(-ENOMEM);
+                    }
+
+                    /* insert right mapping into list */
+                    rm->next = p->next;
+                    p->next = rm;
+
+                    /* update the left mapping length */
+                    p->length = llength;
+                    prev = p;
```

Comment:
We just inserted a new node after p. And `next` is the node after the new node, so `prev` should be the new node, right?

Reply:
Good catch! Fixed. I tested by hand that we hit all case but it still didn't detect this bug.